divviup/src/main/java/org/divviup/android/Client.java

```diff
@@ -0,0 +1,157 @@
+package org.divviup.android;
+
+import org.apache.commons.io.IOUtils;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.URL;
+import java.util.Calendar;
+
+public class Client {
+    private static final String HPKE_CONFIG_LIST_CONTENT_TYPE = "application/dap-hpke-config-list";
+    private static final String REPORT_CONTENT_TYPE = "application/dap-report";
+
+    private static boolean isLibraryLoaded = false;
+
+    private final URI leaderEndpoint, helperEndpoint;
+    private final TaskId taskId;
+    private final long timePrecisionSeconds;
+    private final VDAF vdaf;
+
+
+    private Client(
+            URI leaderEndpoint,
+            URI helperEndpoint,
+            TaskId taskId,
+            long timePrecisionSeconds,
+            VDAF vdaf
+    ) {
+        if (!isLibraryLoaded) {
+            // Load the native library upon first use.
+            System.loadLibrary("divviup_android");
+            isLibraryLoaded = true;
+        }
+
+        if (timePrecisionSeconds <= 0) {
+            throw new IllegalArgumentException("timePrecisionSeconds must be positive");
+        }
+
+        this.leaderEndpoint = leaderEndpoint;
+        this.helperEndpoint = helperEndpoint;
+        this.taskId = taskId;
+        this.timePrecisionSeconds = timePrecisionSeconds;
+        this.vdaf = vdaf;
+    }
+
+    public static Client createPrio3Count(
+            URI leaderEndpoint,
+            URI helperEndpoint,
+            TaskId taskId,
+            long timePrecisionSeconds
+    ) {
+        return new Client(leaderEndpoint, helperEndpoint, taskId, timePrecisionSeconds, VDAF.PRIO3COUNT);
+    }
+
+    public void sendMeasurement(Object measurement) throws IOException {
+        HPKEConfigList leaderConfigList = this.fetchHPKEConfigList(this.leaderEndpoint, this.taskId);
+        HPKEConfigList helperConfigList = this.fetchHPKEConfigList(this.helperEndpoint, this.taskId);
+        long timestamp = this.reportTimestamp();
+        byte[] report = null;
+        if (this.vdaf == VDAF.PRIO3COUNT) {
+            if (measurement instanceof Boolean) {
```

Comment:
Good news: adding a generic type doesn't change the output of `javah`. I think I'm going to keep the `instanceof` checks, because it's possible to shove arbitrary Objects in via raw types.

Reply:
Good idea to keep the `instanceof` checks -- it's unfortunate how "weak" Java's generics are. Still, this is a nicer interface for users of this library.